README.md
[![NPM version](https://img.shields.io/npm/v/excel4node.svg)](https://www.npmjs.com/package/excel4node)
[![License](https://img.shields.io/badge/License-MIT-brightgreen.svg)](https://opensource.org/licenses/MIT)
[![npm](https://img.shields.io/npm/dt/excel4node.svg)](https://www.npmjs.com/package/excel4node)
[![node](https://img.shields.io/node/v/excel4node.svg)](https://nodejs.org/en/download/)
[![Build Status](https://travis-ci.org/natergj/excel4node.svg?branch=master)](https://travis-ci.org/natergj/excel4node)
[![dependencies Status](https://david-dm.org/natergj/excel4node/status.svg)](https://david-dm.org/natergj/excel4node)
[![devDependency Status](https://david-dm.org/natergj/excel4node/dev-status.svg)](https://david-dm.org/natergj/excel4node#info=devDependencies)

# excel4node

A full featured xlsx file generation library allowing for the creation of advanced Excel files.

excel4node conforms to the ECMA-376 OOXML specification 2nd edition

REFERENCES  
[OpenXML White Paper](http://www.ecma-international.org/news/TC45_current_work/OpenXML%20White%20Paper.pdf)  
[ECMA-376 Website](http://www.ecma-international.org/publications/standards/Ecma-376.htm)  
[OpenOffice Excel File Format Reference](http://www.openoffice.org/sc/excelfileformat.pdf)  
[OpenOffice Anatomy of OOXML explanation](http://officeopenxml.com/anatomyofOOXML-xlsx.php)  
[MS-XSLX spec (pdf)](http://download.microsoft.com/download/D/3/3/D334A189-E51B-47FF-B0E8-C0479AFB0E3C/%5BMS-XLSX%5D.pdf)

Code references specifications sections from ECMA-376 2nd edition doc  
ECMA-376, Second Edition, Part 1 - Fundamentals And Markup Language Reference.pdf  
found in ECMA-376 2nd edition Part 1 download at [http://www.ecma-international.org/publications/standards/Ecma-376.htm](http://www.ecma-international.org/publications/standards/Ecma-376.htm)

### Basic Usage

```javascript
// Require library
var xl = require('excel4node');

// Create a new instance of a Workbook class
var wb = new xl.Workbook();

// Add Worksheets to the workbook
var ws = wb.addWorksheet('Sheet 1');
var ws2 = wb.addWorksheet('Sheet 2');

// Create a reusable style
var style = wb.createStyle({
  font: {
    color: '#FF0800',
    size: 12,
  },
  numberFormat: '$#,##0.00; ($#,##0.00); -',
});

// Set value of cell A1 to 100 as a number type styled with paramaters of style
ws.cell(1, 1)
  .number(100)
  .style(style);

// Set value of cell B1 to 200 as a number type styled with paramaters of style
ws.cell(1, 2)
  .number(200)
  .style(style);

// Set value of cell C1 to a formula styled with paramaters of style
ws.cell(1, 3)
  .formula('A1 + B1')
  .style(style);

// Set value of cell A2 to 'string' styled with paramaters of style
ws.cell(2, 1)
  .string('string')
  .style(style);

// Set value of cell A3 to true as a boolean type styled with paramaters of style but with an adjustment to the font size.
ws.cell(3, 1)
  .bool(true)
  .style(style)
  .style({font: {size: 14}});

wb.write('Excel.xlsx');
```

## excel4node

excel4node comes with some generic functions and types

xl.getExcelRowCol(cellRef)  
Accepts cell reference (i.e. 'A1') and returns object with corresponding row and column

```javascript
xl.getExcelRowCol('B5');
// returns { row: 5, col: 2}
```

xl.getExcelAlpha(column)  
Accepts column as integer and returns corresponding column reference as alpha

```javascript
xl.getExcelAlpha(10);
// returns 'J'
```

xl.getExcelCellRef(row, column)  
Accepts row and column as integers and returns Excel cell reference

```javascript
xl.getExcelCellRef(5, 3);
// returns 'C5'
```

xl.getExcelTS(date)  
Accepts Date object and returns an Excel timestamp

```javascript
var newDate = new Date('2015-01-01T00:00:00.0000Z');
xl.getExcelTS(newDate);
// Returns 42004.791666666664
```

xl.PAPER_SIZE

## Workbook

An instance of the Workbook class contains all data and parameters for the Excel Workbook.

#### Constructor

Workbook constructor accepts an optional configuration object.

```javascript
var xl = require('excel4node');
var wb = new xl.Workbook({
  jszip: {
    compression: 'DEFLATE',
  },
  defaultFont: {
    size: 12,
    name: 'Calibri',
    color: 'FFFFFFFF',
  },
  dateFormat: 'm/d/yy hh:mm:ss',
  workbookView: {
    activeTab: 1, // Specifies an unsignedInt that contains the index to the active sheet in this book view.
    autoFilterDateGrouping: true, // Specifies a boolean value that indicates whether to group dates when presenting the user with filtering options in the user interface.
    firstSheet: 1, // Specifies the index to the first sheet in this book view.
    minimized: false, // Specifies a boolean value that indicates whether the workbook window is minimized.
    showHorizontalScroll: true, // Specifies a boolean value that indicates whether to display the horizontal scroll bar in the user interface.
    showSheetTabs: true, // Specifies a boolean value that indicates whether to display the sheet tabs in the user interface.
    showVerticalScroll: true, // Specifies a boolean value that indicates whether to display the vertical scroll bar.
    tabRatio: 600, // Specifies ratio between the workbook tabs bar and the horizontal scroll bar.
    visibility: 'visible', // Specifies visible state of the workbook window. ('hidden', 'veryHidden', 'visible') (§18.18.89)
    windowHeight: 17620, // Specifies the height of the workbook window. The unit of measurement for this value is twips.
    windowWidth: 28800, // Specifies the width of the workbook window. The unit of measurement for this value is twips..
    xWindow: 0, // Specifies the X coordinate for the upper left corner of the workbook window. The unit of measurement for this value is twips.
    yWindow: 440, // Specifies the Y coordinate for the upper left corner of the workbook window. The unit of measurement for this value is twips.
  },
  logLevel: 0, // 0 - 5. 0 suppresses all logs, 1 shows errors only, 5 is for debugging
  author: 'Microsoft Office User', // Name for use in features such as comments
});
```

#### Methods

wb.addWorksheet(name, options);  
Adds a new Worksheet to the Workbook  
Accepts name of new Worksheet and options object (see Worksheet section)  
Returns a Worksheet instance

wb.setSelectedTab(id);  
Sets which tab will be selected when the Workbook is opened  
Accepts Sheet ID (1-indexed sheet in order that sheets were added)

wb.createStyle(opts);  
Creates a new Style instance  
Accepts Style configuration object (see Style section)
Returns a new Style instance

wb.write();  
The write() method can accept a single filename, a filename with callback function or an HTTP response object.

```javascript
var xl = require('excel4node');
var wb = new xl.Workbook();
wb.write('ExcelFile.xlsx'); // Writes the file ExcelFile.xlsx to the process.cwd();
```

```javascript
wb.write('ExcelFile.xlsx', function(err, stats) {
  if (err) {
    console.error(err);
  } else {
    console.log(stats); // Prints out an instance of a node.js fs.Stats object
  }
});
```

```javascript
// sends Excel file to web client requesting the / route
// server will respond with 500 error if excel workbook cannot be generated
var express = require('express');
var app = express();
app.get('/', function(req, res) {
  wb.write('ExcelFile.xlsx', res);
});
app.listen(3000, function() {
  console.log('Example app listening on port 3000!');
});
```

wb.writeToBuffer();  
The writeToBuffer() method access no parameters and returns a promise that resolves with the nodebuffer generated by the JSZip library. This buffer can then be sent to other streams.

```javascript
var xl = require('excel4node');
var wb = new xl.Workbook();

wb.writeToBuffer().then(function(buffer) {
  // Do something with buffer
});
```

## Worksheet

An instance of the Worksheet class contains all information specific to that worksheet

#### Constructor

Worksheet constructor is called via Workbook class and accepts a name and configuration object

```javascript
var xl = require('excel4node');
var wb = new xl.Workbook();

var options = {
  margins: {
    left: 1.5,
    right: 1.5,
  },
};

var ws = wb.addWorksheet('sheetname', options);
```

Full Worksheet options. All options are optional.

```
{
    'margins': { // Accepts a Double in Inches
        'bottom': Double,
        'footer': Double,
        'header': Double,
        'left': Double,
        'right': Double,
        'top': Double
    },
    'printOptions': {
        'centerHorizontal': Boolean,
        'centerVertical': Boolean,
        'printGridLines': Boolean,
        'printHeadings': Boolean

    },
    'headerFooter': { // Set Header and Footer strings and options. See note below
        'evenFooter': String,
        'evenHeader': String,
        'firstFooter': String,
        'firstHeader': String,
        'oddFooter': String,
        'oddHeader': String,
        'alignWithMargins': Boolean,
        'differentFirst': Boolean,
        'differentOddEven': Boolean,
        'scaleWithDoc': Boolean
    },
    'pageSetup': {
        'blackAndWhite': Boolean,
        'cellComments': xl.CellComment, // one of 'none', 'asDisplayed', 'atEnd'
        'copies': Integer,
        'draft': Boolean,
        'errors': xl.PrintError, // One of 'displayed', 'blank', 'dash', 'NA'
        'firstPageNumber': Integer,
        'fitToHeight': Integer, // Number of vertical pages to fit to
        'fitToWidth': Integer, // Number of horizontal pages to fit to
        'horizontalDpi': Integer,
        'orientation': xl.Orientation, // One of 'default', 'portrait', 'landscape'
        'pageOrder': xl.PageOrder, // One of 'downThenOver', 'overThenDown'
        'paperHeight': xl.PositiveUniversalMeasure, // Value must a positive Float immediately followed by unit of measure from list mm, cm, in, pt, pc, pi. i.e. '10.5cm'
        'paperSize': xl.PaperSize, // see lib/types/paperSize.js for all types and descriptions of types. setting paperSize overrides paperHeight and paperWidth settings
        'paperWidth': xl.PositiveUniversalMeasure,
        'scale': Integer,
        'useFirstPageNumber': Boolean,
        'usePrinterDefaults': Boolean,
        'verticalDpi': Integer
    },
    'sheetView': {
        'pane': { // Note. Calling .freeze() on a row or column will adjust these values
            'activePane': xl.Pane, // one of 'bottomLeft', 'bottomRight', 'topLeft', 'topRight'
            'state': xl.PaneState, // one of 'split', 'frozen', 'frozenSplit'
            'topLeftCell': Cell Reference, // i.e. 'A1'

changelog.txt
1.7.2
        Bug Fixes:
        resolves https://github.com/natergj/excel4node/issues/283 to fix issue when comment drawing changing after save
        
1.7.1
        Bug Fixes:
        resolves https://github.com/natergj/excel4node/issues/269 to fix issue where multiple complex strings could not be used
        merges https://github.com/natergj/excel4node/pull/264 to fix README typo
        merges https://github.com/natergj/excel4node/pull/272 to fix license format
        merges https://github.com/natergj/excel4node/pull/275 to address https://github.com/natergj/excel4node/issues/258

1.7.0
        Bug Fixes:
        resolves https://github.com/natergj/excel4node/issues/238 to fix issue with certain emoji characters
        merges https://github.com/natergj/excel4node/pull/240 to fix issue with workbook validation using .Net validation library provided by Microsoft
        resolves https://github.com/natergj/excel4node/issues/236 where first tab will still be shown even if it is hidden

        Enhancements:
        merges https://github.com/natergj/excel4node/pull/243 to add basic cell comment functionality
        
1.6.0
        Bug Fixes:
        merges https://github.com/natergj/excel4node/pull/232 to fix issue where emoji were not being added to worksheets

        Enhancements:
        addresses https://github.com/natergj/excel4node/issues/194 by adding ability to set Print Area for worksheets
        addresses https://github.com/natergj/excel4node/issues/233 by adding ability to set Page Breaks for worksheets

1.5.1
        Bug Fixes:
        resolves https://github.com/natergj/excel4node/issues/218 where library crashed when null or undefined value was sent to cell string method
        resolves https://github.com/natergj/excel4node/issues/226 where default font would not be applied to dates when opening workbook in LibreOffice
        merges https://github.com/natergj/excel4node/pull/225 which removes error when attempting to bundle application with excel4node as dependency

        Enhancements:
        merges https://github.com/natergj/excel4node/pull/230 which reduces library size by only installed specific lodash functions

1.5.0
        Bug Fixes:
        resolves https://github.com/natergj/excel4node/issues/156 where an error will be thrown if no Worksheet is specified prior to attempting to write a Workbook

        Enhancements:
        merge https://github.com/natergj/excel4node/pull/211 & https://github.com/natergj/excel4node/pull/212
                remove default logger dependency in favor of much smaller simple logger (reduces library dependencies and size)
                add ability specify custom logger
1.4.0
        Bug Fixes:
        resolves https://github.com/natergj/excel4node/issues/196 where unicode filenames could not be used with HTTP Response write handler     

        Enhancements:
        merge https://github.com/natergj/excel4node/pull/201 to add ability to hide worksheets
1.3.6
        Bug Fixes:
        resolves https://github.com/natergj/excel4node/issues/182 to allow column widths to be set as floats
        resolves https://github.com/natergj/excel4node/issues/184 to properly handle setting row autofilter with no arguments
        merge https://github.com/natergj/excel4node/pull/181 to fix typo in documentation

1.3.5
        Bug Fixes:
        merge https://github.com/natergj/excel4node/pull/141 to properly handle unicode and emoji
        merge https://github.com/natergj/excel4node/pull/172 to correct issue with row spans causing errors when opening workbooks with a large number of lines
        
1.3.4
        Bug fixes:
        resolved issue where if multiple conditional formatting rules were specified for a single sheet, only the first rule would apply
        resolves https://github.com/natergj/excel4node/issues/164 to resolve issue adding multiple dataValidations that did not include a formula to a single worksheet

        Enhancements:
        merge https://github.com/natergj/excel4node/pull/165 to improve performance with shared strings
        
1.3.2
        Bug fixes:
        resolves https://github.com/natergj/excel4node/issues/152 after 1.3.0 re-introduced issue #84
        
1.3.1
        Bug fixes:
        resolves https://github.com/natergj/excel4node/issues/148 for uncatchable thrown errors
        resolves https://github.com/natergj/excel4node/issues/106 for incorrectly generated links
        resolves https://github.com/natergj/excel4node/issues/147 for missing fills in conditional formatting

1.3.0
        Enhancements:
        merge https://github.com/natergj/excel4node/pull/117 to add option to hide view lines
        merge https://github.com/natergj/excel4node/pull/109 to add code coverage checking
        merge https://github.com/natergj/excel4node/pull/138 to allow for image from buffer support
        commit 1c4a7bf990160b47ad8e0f49c00a536dca7ab672 add options for workbook view
        commit ff629a225335dc784933c06aabd20ddca5fcdbef fix issue where ony first data validation per sheet would take effect

        Bug fixes:
        merge https://github.com/natergj/excel4node/pull/99 to fix issue when adding 2nd image of same extension

sample.js
const fs = require('fs');
const path = require('path');

require('source-map-support').install();
var xl = require('./distribution');

function generateWorkbook() {
    var wb = new xl.Workbook({
        defaultFont: {
            name: 'Verdana',
            size: 12
        },
        dateFormat: 'mm/dd/yyyy hh:mm:ss',
        logLevel: 1,
    });

    /*****************************************
     * START Create a sample invoice
     *****************************************/

    // Create some styles to be used throughout
    var multiLineStyle = wb.createStyle({
        alignment: {
            wrapText: true,
            vertical: 'top'
        }
    });
    var largeText = wb.createStyle({
        font: {
            name: 'Cambria',
            size: 20
        }
    });
    var medText = wb.createStyle({
        font: {
            name: 'Cambria',
            size: 14,
            color: '#D4762C'
        },
        alignment: {
            vertical: 'center'
        }
    });

    var currencyStyle = wb.createStyle({
        numberFormat: '$##0.00; [Red]($##0.00); $0.00'
    });

    var invoiceWS = wb.addWorksheet('Invoice', {
        pageSetup: {
            fitToWidth: 1
        },
        headerFooter: {
            oddHeader: 'iAmNater invoice',
            oddFooter: 'Invoice Page &P'
        }
    });

    // Set some row and column properties
    invoiceWS.row(1).setHeight(25);
    invoiceWS.row(2).setHeight(45);
    invoiceWS.column(1).setWidth(3);
    invoiceWS.column(2).setWidth(10);
    invoiceWS.column(3).setWidth(35);
    invoiceWS.column(5).setWidth(25);
    invoiceWS.cell(2, 2).string('INVOICE').style(largeText);
    invoiceWS.cell(2, 3).string('809871').style(largeText).style({
        font: {
            color: '#D4762C'
        }
    });

    // Add a company logo
    invoiceWS.addImage({
        path: './sampleFiles/logo.png',
        type: 'picture',
        position: {
            type: 'twoCellAnchor',
            from: {
                col: 4,
                colOff: 0,
                row: 2,
                rowOff: 0
            },
            to: {
                col: 6,
                colOff: 0,
                row: 3,
                rowOff: 0
            }
        }
    });

    // Add some borders to specific cells
    invoiceWS.cell(2, 2, 2, 5).style({
        border: {
            bottom: {
                style: 'thick',
                color: '#000000'
            }
        }
    });

    // Add some data and adjust styles for specific cells
    invoiceWS.cell(3, 2, 3, 3, true).string('January 1, 2016').style({
        border: {
            bottom: {
                style: 'thin',
                color: '#D4762C'
            }
        }
    });
    invoiceWS.cell(4, 2, 4, 3, true).string('PAYMENT DUE BY: March 1, 2016').style({
        font: {
            bold: true
        }
    });

    // style methods can be chained. multiple styles will be merged with last style taking precedence if there is a conflict
    invoiceWS.cell(3, 5, 4, 5, true).formula('E31').style(currencyStyle).style({
        font: {
            size: 20,
            color: '#D4762C'
        },
        alignment: {
            vertical: 'center'
        }
    });
    invoiceWS.cell(4, 2, 4, 5).style({
        border: {
            bottom: {
                style: 'thin',
                color: '#000000'
            }
        }
    });

    invoiceWS.row(6).setHeight(75);
    invoiceWS.cell(6, 2, 6, 5).style(multiLineStyle);

    // set some strings to have multiple font formats within a single cell
    invoiceWS.cell(6, 2, 6, 3, true).string([{
            bold: true
        },
        'Client Name\n',
        {
            bold: false
        },
        'Company Name Inc.\n1234 First Street\nSomewhere, OR 12345'
    ]);

    invoiceWS.cell(6, 4, 6, 5, true).string([{
            bold: true
        },
        'iAmNater.com\n',
        {
            bold: false
        },
        '123 Nowhere Lane\nSomewhere, OR 12345'
    ]).style({
        alignment: {
            horizontal: 'right'
        }
    });

    invoiceWS.cell(8, 2, 8, 5).style({
        border: {
            bottom: {
                style: 'thick',
                color: '#000000'
            }
        }
    });

    invoiceWS.cell(10, 2).string('QUANTITY');
    invoiceWS.cell(10, 3).string('DETAILS');
    invoiceWS.cell(10, 4).string('UNIT PRICE').style({
        alignment: {
            horizontal: 'right'
        }
    });
    invoiceWS.cell(10, 5).string('LINE TOTAL').style({
        alignment: {
            horizontal: 'right'
        }
    });

    var items = require('./sampleFiles/invoiceData.json').items;
    var i = 0;
    var rowOffset = 11;
    var oddBackgroundColor = '#F8F5EE';
    while (i <= 10) {
        var item = items[i];
        var curRow = rowOffset + i;
        if (item !== undefined) {
            invoiceWS.cell(curRow, 2).number(item.units).style({
                alignment: {
                    horizontal: 'left'
                }
            });
            invoiceWS.cell(curRow, 3).string(item.description);
            invoiceWS.cell(curRow, 4).number(item.unitCost).style(currencyStyle);
            invoiceWS.cell(curRow, 5).formula(xl.getExcelCellRef(rowOffset + i, 2) + '*' + xl.getExcelCellRef(rowOffset + 1, 4)).style(currencyStyle);
        }
        if (i % 2 === 0) {
            invoiceWS.cell(curRow, 2, curRow, 5).style({
                fill: {
                    type: 'pattern',
                    patternType: 'solid',
                    fgColor: oddBackgroundColor
                }
            });
        }
        i++;
    }
    invoiceWS.cell(21, 2, 21, 5).style({
        border: {
            bottom: {
                style: 'thin',
                color: '#DCD1B3'
            }
        }
    });

    invoiceWS.cell(22, 4).string('Discount');
    invoiceWS.cell(22, 5).number(0.00).style(currencyStyle);

    invoiceWS.cell(23, 4).string('Net Total');
    invoiceWS.cell(23, 5).formula('SUM(E11:E21)').style(currencyStyle);

    invoiceWS.cell(23, 2, 23, 5).style({
        border: {
            bottom: {
                style: 'thin',
                color: '#000000'
            }
        }
    });

    invoiceWS.row(24).setHeight(20);
    invoiceWS.cell(24, 4, 25, 4, true).string('USD TOTAL').style(medText);
    invoiceWS.cell(24, 5, 25, 5, true).formula('SUM(E22:E23)').style(medText).style(currencyStyle);
    /*****************************************
     * END Create a sample invoice
     *****************************************/


    /*****************************************
     * START Create a filterable list
     *****************************************/

    var filterSheet = wb.addWorksheet('Filters', {
        disableRowSpansOptimization: true,
    });

    for (var i = 1; i <= 10; i++) {
        filterSheet.cell(1, i).string('Header' + i);
    }
    filterSheet.row(1).filter(1, 10);

    for (var r = 2; r <= 30; r++) {
        for (var c = 1; c <= 10; c++) {
            filterSheet.cell(r, c).number(parseInt(Math.random() * 100));
        }
    }
    /*****************************************
     * END Create a filterable list
     *****************************************/

    /*****************************************
     * START Create collapsable lists
     *****************************************/


    var collapseSheet = wb.addWorksheet('Collapsables', {
        pageSetup: {
            fitToWidth: 1
        },
        outline: {
            summaryBelow: true
        }
    });

    var rowOffset = 0;
    for (var r = 1; r <= 10; r++) {
        for (var c = 1; c <= 10; c++) {
            collapseSheet.cell(r + rowOffset, c).number(parseInt(Math.random() * 100));
        }
        collapseSheet.row(r + rowOffset).group(1, true);
    }
    for (var i = 1; i <= 10; i++) {
        collapseSheet.cell(11, i).formula('SUM(' + xl.getExcelCellRef(rowOffset + 1, i) + ':' + xl.getExcelCellRef(rowOffset + 10, i) + ')');
    }
    collapseSheet.cell(11, 1, 11, 10).style({
        fill: {
            type: 'pattern',
            patternType: 'solid',
            fgColor: '#C2D6EC'
        }
    });

    var rowOffset = 11;
    for (var r = 1; r <= 10; r++) {
        for (var c = 1; c <= 10; c++) {
            collapseSheet.cell(r + rowOffset, c).number(parseInt(Math.random() * 100));
        }
        collapseSheet.row(r + rowOffset).group(1, true);
    }
    for (var i = 1; i <= 10; i++) {
        collapseSheet.cell(22, i).formula('SUM(' + xl.getExcelCellRef(rowOffset + 1, i) + ':' + xl.getExcelCellRef(rowOffset + 10, i) + ')');
    }
    collapseSheet.cell(22, 1, 22, 10).style({
        fill: {
            type: 'pattern',
            patternType: 'solid',
            fgColor: '#4273B0'
        }
    });



    var rowOffset = 22;
    for (var r = 1; r <= 10; r++) {
        for (var c = 1; c <= 10; c++) {
            collapseSheet.cell(r + rowOffset, c).number(parseInt(Math.random() * 100));
        }
        collapseSheet.row(r + rowOffset).group(1, true);
    }
    for (var i = 1; i <= 10; i++) {
        collapseSheet.cell(33, i).formula('SUM(' + xl.getExcelCellRef(rowOffset + 1, i) + ':' + xl.getExcelCellRef(rowOffset + 10, i) + ')');
    }
    collapseSheet.cell(33, 1, 33, 10).style({
        fill: {
            type: 'pattern',
            patternType: 'solid',
            fgColor: '#C2D6EC'
        }
    });

    var rowOffset = 33;
    for (var r = 1; r <= 10; r++) {
        for (var c = 1; c <= 10; c++) {
            collapseSheet.cell(r + rowOffset, c).number(parseInt(Math.random() * 100));
        }
        collapseSheet.row(r + rowOffset).group(1, true);
    }
    for (var i = 1; i <= 10; i++) {
        collapseSheet.cell(44, i).formula('SUM(' + xl.getExcelCellRef(rowOffset + 1, i) + ':' + xl.getExcelCellRef(rowOffset + 10, i) + ')');
    }
    collapseSheet.cell(44, 1, 44, 10).style({
        fill: {
            type: 'pattern',
            patternType: 'solid',
            fgColor: '#4273B0'
        }
    });



    var rowOffset = 44;
    for (var r = 1; r <= 10; r++) {
        for (var c = 1; c <= 10; c++) {
            collapseSheet.cell(r + rowOffset, c).number(parseInt(Math.random() * 100));
        }
        collapseSheet.row(r + rowOffset).group(1, true);
    }
    for (var i = 1; i <= 10; i++) {
        collapseSheet.cell(55, i).formula('SUM(' + xl.getExcelCellRef(rowOffset + 1, i) + ':' + xl.getExcelCellRef(rowOffset + 10, i) + ')');
    }
    collapseSheet.cell(55, 1, 55, 10).style({
        fill: {
            type: 'pattern',
            patternType: 'solid',
            fgColor: '#C2D6EC'
        }
    });

    var rowOffset = 55;
    for (var r = 1; r <= 10; r++) {
        for (var c = 1; c <= 10; c++) {
            collapseSheet.cell(r + rowOffset, c).number(parseInt(Math.random() * 100));
        }
        collapseSheet.row(r + rowOffset).group(1, true);
    }
    for (var i = 1; i <= 10; i++) {
        collapseSheet.cell(66, i).formula('SUM(' + xl.getExcelCellRef(rowOffset + 1, i) + ':' + xl.getExcelCellRef(rowOffset + 10, i) + ')');
    }
    collapseSheet.cell(66, 1, 66, 10).style({
        fill: {
            type: 'pattern',
            patternType: 'solid',
            fgColor: '#4273B0'
        }
    });
    /*****************************************
     * START Create collapsable lists
     *****************************************/

    /*****************************************
     * START Create Frozen lists
     *****************************************/

    var frozenSheet = wb.addWorksheet('Frozen');

    for (var i = 2; i <= 21; i++) {
        frozenSheet.cell(1, i).string('Column' + i);
    }
    frozenSheet.row(1).freeze();

    for (var r = 2; r <= 30; r++) {
        frozenSheet.cell(r, 1).string('Row' + r);
        for (var c = 2; c <= 22; c++) {
            frozenSheet.cell(r, c).number(parseInt(Math.random() * 100));
        }
    }
    frozenSheet.column(1).freeze();

    /*****************************************
     * END Create Frozen lists
     *****************************************/

    /*****************************************
     * START Create Split sheet
     *****************************************/

    var splitSheet = wb.addWorksheet('SplitSheet', {
        'sheetView': {
            'pane': {
                'activePane': 'bottomRight',
                'state': 'split',
                'xSplit': 2000,
                'ySplit': 3000
            }
        }
    });

    for (var r = 1; r <= 30; r++) {
        for (var c = 1; c <= 20; c++) {
            splitSheet.cell(r, c).number(parseInt(Math.random() * 100));
        }
    }

    /*****************************************
     * END Create Split
     *****************************************/

    /*****************************************
     * START Create Selectable Options list
     *****************************************/
    var optionsSheet = wb.addWorksheet('Selectable Options');

    optionsSheet.cell(1, 1).string('Booleans');
    optionsSheet.cell(1, 2).string('Option List');
    optionsSheet.cell(1, 3).string('Numbers 1-10');

    optionsSheet.addDataValidation({
        type: 'list',
        allowBlank: true,
        prompt: 'Choose from dropdown',
        error: 'Invalid choice was chosen',
        sqref: 'A2:A10',
        formulas: [
            'true,false'
        ]
    });

    optionsSheet.addDataValidation({
        type: 'list',
        allowBlank: true,
        prompt: 'Choose from dropdown',
        promptTitle: 'Choose from dropdown',
        error: 'Invalid choice was chosen',
        showInputMessage: true,
        sqref: 'B2:B10',
        formulas: [
            'option 1,option 2,option 3'
        ]
    });

    optionsSheet.addDataValidation({
        errorStyle: 'stop',
        error: 'Number must be between 1 and 10',
        type: 'whole',
        operator: 'between',
        allowBlank: 1,
        sqref: 'C2:C10',
        formulas: [1, 10]
    });
    /*****************************************
     * END Create Selectable Options list
     *****************************************/

    /*****************************************
     * START date check sheet
     *****************************************/

    var datesSheet = wb.addWorksheet('dates');

    datesSheet.column(1).setWidth(24);
    datesSheet.column(2).setWidth(20);
    datesSheet.column(3).setWidth(45);

    datesSheet.cell(1, 1).string('Dates');
    datesSheet.cell(1, 2).string('Serial Value');
    datesSheet.cell(1, 3).string('Date String');

    datesSheet.cell(2, 1).date(new Date('1910-02-03T10:05:54Z'));
    datesSheet.cell(2, 2).string('3687.4207639');
    datesSheet.cell(2, 3).string('Correctly translated 1910-02-03T10:05:54Z');
    datesSheet.cell(3, 1).date(new Date('1900-01-01T12:00:00Z'));
    datesSheet.cell(3, 2).string('1.5000000');
    datesSheet.cell(3, 3).string('Correctly translated 1900-01-01T12:00:00Z');
    datesSheet.cell(4, 1).date(new Date('9999-12-31T23:59:59Z'));
    datesSheet.cell(4, 2).string('2958465.9999884');
    datesSheet.cell(4, 3).string('Correctly translated 9999-12-31T23:59:59Z');


    datesSheet.cell(5, 1).date(new Date('1900-01-01T00:00:00Z'));
    datesSheet.cell(5, 2).string('1');
    datesSheet.cell(5, 3).string('Correctly translated 1900-01-01');
    datesSheet.cell(6, 1).date(new Date('1910-02-03T00:00:00Z'));
    datesSheet.cell(6, 2).string('3687');
    datesSheet.cell(6, 3).string('Correctly translated 1910-02-03');
    datesSheet.cell(7, 1).date(new Date('2006-02-01T00:00:00Z'))
    datesSheet.cell(7, 2).string('38749');
    datesSheet.cell(7, 3).string('Correctly translated 2006-02-01');
    datesSheet.cell(8, 1).date(new Date('9999-12-31T00:00:00Z'));
    datesSheet.cell(8, 2).string('2958465');
    datesSheet.cell(8, 3).string('Correctly translated 9999-12-31');


    /*****************************************
     * END date check sheet
     *****************************************/

    /*****************************************
     * START valid characters
     *****************************************/
    var unicodeSheet = wb.addWorksheet('unicode');

    unicodeSheet.column(1).setWidth(40);
    unicodeSheet.column(2).setWidth(55);
    unicodeSheet.column(3).setWidth(65);

    unicodeSheet.cell(1, 1).string('Unicode String');
    unicodeSheet.cell(1, 2).string('Text Representation');
    unicodeSheet.cell(1, 3).string('Encoded Representation');

    unicodeSheet.cell(2, 1).string('Hi <>');
    unicodeSheet.cell(2, 2).string('Hi [less than][greater than]');
    unicodeSheet.cell(2, 3).string('Hi <>');

    unicodeSheet.cell(3, 1).string('😂');
    unicodeSheet.cell(3, 2).string('[face with tears of joy]');
    unicodeSheet.cell(3, 3).string('&#x1f602;');

    unicodeSheet.cell(4, 1).string('hello! 😂');
    unicodeSheet.cell(4, 2).string('hello! [face with tears of joy]');
    unicodeSheet.cell(4, 3).string('hello! &#x1f602;');

    unicodeSheet.cell(5, 1).string('☕️');
    unicodeSheet.cell(5, 2).string('[hot beverage]');
    unicodeSheet.cell(5, 3).string('☕️ (not escaped)');

    unicodeSheet.cell(6, 1).string('😂☕️');
    unicodeSheet.cell(6, 2).string('[face with tears of joy][hot beverage]');
    unicodeSheet.cell(6, 3).string('&#x1f602;☕️ (not eescaped)');

    unicodeSheet.cell(7, 1).string('Good 🤞🏼 Luck');
    unicodeSheet.cell(7, 2).string('Good [crossed fingers: light skin tone] luck');
    unicodeSheet.cell(7, 3).string('Good &#x1f91e;&#x1f3fc; Luck');

    unicodeSheet.cell(8, 1).string('Fist 🤜🏻🤛🏿 bump');
    unicodeSheet.cell(8, 2).string('Fist [right-facing fist: light skin tone][left-facing fist: dark skin tone] bump');
    unicodeSheet.cell(8, 3).string('Fist &#x1f91c;&#x1f3fb;&#x1f91b;&#x1f3ff; bump');

    unicodeSheet.cell(9, 1).string('I am the Α and the Ω');
    unicodeSheet.cell(9, 2).string('I am the [greek capital letter alpha] and the [greek capital letter omega]');
    unicodeSheet.cell(9, 3).string('I am the Α and the Ω (not escaped)');

    unicodeSheet.cell(10, 1).string('𐤶');
    unicodeSheet.cell(10, 2).string('[lydian letter En]');
    unicodeSheet.cell(10, 3).string('&#x20046;');

    unicodeSheet.cell(11, 1).string('𠁆');
    unicodeSheet.cell(11, 2).string('[deograph bik6]');
    unicodeSheet.cell(11, 3).string('&#x10936;');

    unicodeSheet.cell(11, 1).string(String.fromCodePoint(0x000b));
    unicodeSheet.cell(11, 2).string('[vertical tab]');
    unicodeSheet.cell(11, 3).string('');
    /*****************************************
     * END valid characters
     *****************************************/

    /*****************************************
     * START final sheet
     *****************************************/

    var imageSheet = wb.addWorksheet('images', {
        pageSetup: {
            orientation: 'landscape'
        },
        sheetView: {
            zoomScale: 120
        }
    });

    imageSheet.cell(1, 1).string('Images');
    imageSheet.addImage({
        path: './sampleFiles/thumbsUp.jpg',
        type: 'picture',
        position: {
            type: 'absoluteAnchor',
            x: 0,
            y: '10mm'
        }
    });
    imageSheet.cell(30, 1).string('👍');
    imageSheet.addImage({
        image: fs.readFileSync(path.resolve(__dirname, './sampleFiles/thumbsUp.jpg')),
        type: 'picture',
        position: {
            type: 'oneCellAnchor',
            from: {
                col: 1,
                colOff: 0,
                row: 32,
                rowOff: 0
            }
        }
    });

    /*****************************************
     * END final sheet
     *****************************************/

    return wb;
}


var wb = generateWorkbook();

/*
wb.write('Excel1.xlsx');
*/

wb.write('Excel.xlsx', function (err, stats) {
    if (err) {
        console.log(err);
    } else {
        console.log('Excel.xlsx written and has the following stats');
        console.log(stats);
    }
});

/*
var http = require('http');
http.createServer(function (req, res) {
    wb.write('MyExcel.xlsx', res);
}).listen(3000, function () {
    console.log('Go to http://localhost:3000 to download a copy of this Workbook');
});
*/